gateway/src/apicast/policy_schemas_loader.lua

```diff
@@ -0,0 +1,99 @@
+--- Policy schemas loader
+-- Finds the schemas for the builtin policies and the other ones loaded.
+
+local pl_file = require('pl.file')
+local pl_dir = require('pl.dir')
+local pl_path = require('pl.path')
+local cjson = require('cjson')
+local format = string.format
+local ipairs = ipairs
+local insert = table.insert
+local policy_loader = require('apicast.policy_loader')
+
+local _M = {}
+
+local builtin_policy_load_path = policy_loader.builtin_policy_load_path
+local policy_load_paths = policy_loader.policy_load_paths
+
+local policy_schema_name = 'apicast-policy.json'
+
+-- Returns the schemas for all the built-in policies.
+local function all_builtin_policy_schemas()
+  local schemas = {}
+
+  local builtin_policy_dirs = pl_dir.getdirectories(builtin_policy_load_path)
+
+  for _, policy_dir in ipairs(builtin_policy_dirs) do
+    local schema_file = format('%s/%s', policy_dir, policy_schema_name)
+    local schema = pl_file.read(schema_file)
+    if schema then insert(schemas, cjson.decode(schema)) end
+  end
+
+  return schemas
+end
+
+-- Returns a schema from a 'version' directory. Policies that are not built-in,
+-- are located in a path like policies/my_policy/1.0.0.
+-- This function tries to fetch a schema starting from that `1.0.0` directory.
+-- It looks for the json schema, gets its version, and compares it against the
+-- version in the directory.
+-- Returns the schema when found. When it's not present or there's a version
+-- mismatch, it returns nil.
+local function get_schema_from_version_dir(version_dir)
+  local schema_file = format('%s/%s', version_dir, policy_schema_name)
+  local schema = pl_file.read(schema_file)
+
+  if schema then
+    local decoded_schema = cjson.decode(schema)
+    local version_in_schema = decoded_schema.version
+    local version_in_path = pl_path.basename(version_dir)
+
+    if version_in_path == version_in_schema then
+      return decoded_schema
+    else
+      ngx.log(ngx.WARN,
+        'Could not load ', decoded_schema.name,
+        ' version in schema is ', version_in_schema,
+        ' but version in path is ', version_in_path)
+    end
+  end
+end
+
+-- Returns all the non-built-in schemas from the paths specified by the user.
+-- These paths always follow the same pattern. It contains a directory for each
+-- policy, and each of those contain a directory for each version of that
+-- policy. The json schema is in that 'version' directory.
+local function all_loaded_policy_schemas()
+  local schemas = {}
+
+  for _, load_path in ipairs(policy_load_paths) do
+    if pl_path.exists(load_path) then
+      local policy_dirs = pl_dir.getdirectories(load_path)
+      for _, policy_dir in ipairs(policy_dirs) do
+        local version_dirs = pl_dir.getdirectories(policy_dir)
+        for _, version_dir in ipairs(version_dirs) do
```

Comment:
I think it could be a bit nicer if we would reduce the number of variables by defining something like `dir_iter(dir)` that would do `return ipairs(pl_dir.getdirectories(dir))`

Reply:
Thanks. I changed it. Looks better :ok_hand: